internal/pkg/policy/policy_output.go

```diff
@@ -56,18 +56,73 @@ func (p *PolicyOutput) Prepare(ctx context.Context, zlog zerolog.Logger, bulker
 		// Note: This will need to be updated when doing multi-cluster elasticsearch support
 		// Currently, we only have access to the token for the elasticsearch instance fleet-server
 		// is monitors. When updating for multiple ES instances we need to tie the token to the output.
-		needNewKey := true
+		needNewKey := false
+		needUpdateKey := false
 		switch {
 		case agent.DefaultAPIKey == "":
 			zlog.Debug().Msg("must generate api key as default API key is not present")
+			needNewKey = true
 		case p.Role.Sha2 != agent.PolicyOutputPermissionsHash:
 			zlog.Debug().Msg("must generate api key as policy output permissions changed")
+			needUpdateKey = true
 		default:
-			needNewKey = false
 			zlog.Debug().Msg("policy output permissions are the same")
 		}
 
-		if needNewKey {
+		if needUpdateKey {
+			zlog.Debug().
+				RawJSON("roles", p.Role.Raw).
+				Str("oldHash", agent.PolicyOutputPermissionsHash).
+				Str("newHash", p.Role.Sha2).
+				Msg("Generating a new API key")
+
+			// query current api key for roles so we don't lose permissions in the meantime
+			currentRoles, err := fetchAPIKeyRoles(ctx, bulker, agent.DefaultAPIKeyID)
+			if err != nil {
+				zlog.Error().
+					Str("apiKeyID", agent.DefaultAPIKeyID).
+					Err(err).Msg("fail fetching roles for key")
+				return err
+			}
+
+			// merge roles with p.Role
+			newRoles, err := mergeRoles(zlog, currentRoles, p.Role)
+			if err != nil {
+				zlog.Error().
+					Str("apiKeyID", agent.DefaultAPIKeyID).
+					Err(err).Msg("fail merging roles for key")
+				return err
+			}
+
+			// hash provided is only for merging request together and not persisted
+			err = bulker.APIKeyUpdate(ctx, agent.DefaultAPIKeyID, newRoles.Sha2, newRoles.Raw)
+			if err != nil {
+				zlog.Error().Err(err).Msg("fail generate output key")
+				zlog.Debug().RawJSON("roles", newRoles.Raw).Str("sha", newRoles.Sha2).Err(err).Msg("fail generate output key")
```

Comment:
[Suggestion]
Please, make the messages different, it'd be confusing seeing the same message twice with different levels